src/jquery.interactive-world-map.js

```diff
@@ -1,142 +1,152 @@
+    var settings;
+    var dataCountries = {};                 //Hold the countries information sent by parameters
+    var countriesFill = {};                 //Hold the countries code that must be Highlighted for be a YFU Country
+    var hoverPattern;                       //The svg pattern (html code) that uses the hover colors over the YFU countries and make it dotted.
+    var yfuCountriesHighlighted;            //The pattern that makes the yfu countries dotted and highlighted
+    var generalDottedPatternStyle;          //The general dotted style for all the countries
+
+    var defaults = {
+        colors: {
+            general: '#cccccc',
+            hover: 'red',
+            highlighted: 'green'
+        },
+        popup: {
+            margins: {top: 100, left: 60}
+        },
+        svg: {
+           pattern: {
+                circle: {
+                    x: 1, y: 1, cx:3, cy: 3, r: 3
+                },
+               size: {
+                   width:  "5",
+                   heigth: "5"
+               }
+           }
+        }
+    };
 
-            $data = $(this).data();
 
-            if($data.margincountrypopupfortop) {
-                marginCountryPopupForTop            = $data.margincountrypopupfortop;
-            }
+    function interactiveWorldMap(element, options) {
+        this.element    = element;
+        this._options   = options;
+        this.init();
+    }
 
-            if($data.margincountrypopupforleft) {
-                marginCountryPopupForLeft          = $data.margincountrypopupforleft;
-            }
 
-            if($data.colorhover){
-                $.fn.interactiveWorldMap.defaults.hover       = $data.colorhover
-            }
-            if($data.colorhighlighted){
-                $.fn.interactiveWorldMap.defaults.highlighted = $data.colorhighlighted;
+    $.extend(interactiveWorldMap.prototype, {
+            init: function () {
+                settings   = $.extend(defaults, $(this.element).data(), this._options);
+                if ($.isPlainObject(settings.countries)) {
+                    dataCountries = settings.countries;
+                }else {
+                    dataCountries = jQuery.parseJSON(settings.countries);
+                }
+                configMapStyle(this.element);                //Configurations for the map. Dotted Style and popup prepended
             }
         }
+    );
 
-        var countriesFill = {};
-        $.each(dataCountries, function(index, value) {
-                countriesFill[index] = 'url(#dotsHighlighted)';
-            }
-        );
+    function configMapStyle(element) {
+        generalDottedPatternStyle = createPatternSvg("dotsGeneral",     settings.colors.general);
+        yfuCountriesHighlighted   = createPatternSvg("dotsHighlighted", settings.colors.highlighted);
+        hoverPattern              = createPatternSvg("dotsHover",       settings.colors.hover);
+        prependModalContainer(element);
```

Comment:
The problem I have in that it's when  you call the plugin like this:

$.fn[ pluginName ] = function(options) {
    return this.each(function() {
            //Call the constructor
            new interactiveWorldMap(this, options)

Here when I call the plugin  interactiveWorldMap with the div I setted in
the index.html,  I send "this" for use it in the methods outside the
plugin. And f.e If I do console.log(this.element) without send the
parameter in configMapStyle() it's empty

2015-10-19 18:43 GMT+02:00 icatcher.at <notifications@github.com>:

> In src/jquery.interactive-world-map.js
> <https://github.com/yfu-is/interactive-world-map/pull/11#discussion_r42394925>
> :
>
> >
> > -        var countriesFill = {};
> > -        $.each(dataCountries, function(index, value) {
> > -                countriesFill[index] = 'url(#dotsHighlighted)';
> > -            }
> > -        );
> > +    function configMapStyle(element) {
> > +        generalDottedPatternStyle = createPatternSvg("dotsGeneral",     settings.colors.general);
> > +        yfuCountriesHighlighted   = createPatternSvg("dotsHighlighted", settings.colors.highlighted);
> > +        hoverPattern              = createPatternSvg("dotsHover",       settings.colors.hover);
> > +        prependModalContainer(element);
>
> Do you really need to pass element as a parameter? It's in the class
> context with this, no?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yfu-is/interactive-world-map/pull/11/files#r42394925>.
>